R/nn.R

```diff
@@ -637,6 +637,56 @@ length.nn_sequential <- function(x) {
   nn_sequential(!!!lapply(y, function(i) x[[i]]))
 }
 
+#' Prune top layer(s) of a network
+#'
+#' Prune `head_size` last layers of a nn_module in order to 
+#'  replace them by your own head, or in order to use the pruned module
+#'  as a sequential embedding module.
+#' @param x nn_network to prune
+#' @param head_size number of nn_layers to prune 
+#'
+#' @return a nn_sequential network with the top nn_layer removed
+#' @export
+#'
+#' @examples
+#' 
+#' # Remove the classification head of a resnet34 pretrained network
+#' if (torch_is_installed()) {
+#'   resnet34 <- torchvision::network_resnet34(pretrained=TRUE)
+#'   resnet34_headless <- nn_prune_head(resnet34)
+#' }
+nn_prune_head <- function(x, head_size) {
+  UseMethod("nn_prune_head")
+}
+
+#' @export
+nn_prune_head.nn_sequential <- function(x, head_size=1L ) {
+  nn_sequential(!!!x$children[1:(length(x)-head_size)])
+}
+
+#' @export
+nn_prune_head.nn_module <- module <- nn_module(
+    classname = "nn_sequential",
+    initialize = function(x, head_size=1L) {
+      modules <- rlang::list2(!!!x$children[1:(length(x$children)-head_size)])
+      mod_names <- names(modules) 
+      for (i in seq_along(modules)) {
+        self$add_module(name = mod_names[i], module = modules[[i]])
+      }
+    },
+    forward = function(...) {
+      i <- 1
```

Comment:
I don't see where `i` is updated here. Shouldn't it change in the loop?

Reply:
you're absolutely right. my bad, sorry. I've turned it into an explicit flag.